File: backend/app/modules/fleet/api/fuel_pump_deliveries.py
"""Fuel pump delivery endpoints."""

from datetime import date

from fastapi import APIRouter, Query

from app.shared.api.deps import CurrentUserDep, DBDep
from app.schemas import (
    FuelPumpDeliveryCreate,
    FuelPumpDeliveryResponse,
    FuelPumpDeliveryUpdate,
)
from app.services import fuel_pump_delivery_service

router = APIRouter()


@router.get("", response_model=list[FuelPumpDeliveryResponse])
def list_fuel_pump_deliveries(
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
    pump_id: str | None = None,
    start_date: date | None = None,
    end_date: date | None = None,
) -> list[FuelPumpDeliveryResponse]:
    """List fuel pump deliveries with optional filters."""
    if pump_id:
        deliveries = fuel_pump_delivery_service.get_by_pump(
            db,
            pump_id=pump_id,
            skip=skip,
            limit=limit,
        )
    elif start_date and end_date:
        deliveries = fuel_pump_delivery_service.get_by_date_range(
            db,
            organization_id=current_user.organization_id,
            start_date=start_date,
            end_date=end_date,
            skip=skip,
            limit=limit,
        )
    else:
        deliveries = fuel_pump_delivery_service.get_multi(
            db,
            skip=skip,
            limit=limit,
            organization_id=current_user.organization_id,
        )
    return [FuelPumpDeliveryResponse.model_validate(d) for d in deliveries]


@router.post("", response_model=FuelPumpDeliveryResponse, status_code=201)
def create_fuel_pump_delivery(
    delivery_in: FuelPumpDeliveryCreate,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpDeliveryResponse:
    """Create a new fuel pump delivery (automatically updates pump level)."""
    delivery_data = delivery_in.model_copy(
        update={"organization_id": current_user.organization_id}
    )
    delivery = fuel_pump_delivery_service.create(db, obj_in=delivery_data)
    return FuelPumpDeliveryResponse.model_validate(delivery)


@router.get("/summary")
def get_delivery_summary(
    db: DBDep,
    current_user: CurrentUserDep,
    start_date: date = Query(..., description="Start date for summary"),
    end_date: date = Query(..., description="End date for summary"),
) -> dict:
    """Get delivery summary for a date range."""
    return fuel_pump_delivery_service.get_summary(
        db,
        organization_id=current_user.organization_id,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/pump/{pump_id}", response_model=list[FuelPumpDeliveryResponse])
def list_deliveries_by_pump(
    pump_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
) -> list[FuelPumpDeliveryResponse]:
    """List all deliveries for a specific pump."""
    deliveries = fuel_pump_delivery_service.get_by_pump(
        db,
        pump_id=pump_id,
        skip=skip,
        limit=limit,
    )
    return [FuelPumpDeliveryResponse.model_validate(d) for d in deliveries]


@router.get("/{delivery_id}", response_model=FuelPumpDeliveryResponse)
def get_fuel_pump_delivery(
    delivery_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpDeliveryResponse:
    """Get a fuel pump delivery by ID."""
    delivery = fuel_pump_delivery_service.get_or_404(db, delivery_id)
    return FuelPumpDeliveryResponse.model_validate(delivery)


@router.patch("/{delivery_id}", response_model=FuelPumpDeliveryResponse)
def update_fuel_pump_delivery(
    delivery_id: str,
    delivery_in: FuelPumpDeliveryUpdate,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpDeliveryResponse:
    """Update a fuel pump delivery."""
    delivery = fuel_pump_delivery_service.get_or_404(db, delivery_id)
    updated = fuel_pump_delivery_service.update(db, db_obj=delivery, obj_in=delivery_in)
    return FuelPumpDeliveryResponse.model_validate(updated)


@router.delete("/{delivery_id}", status_code=204)
def delete_fuel_pump_delivery(
    delivery_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
) -> None:
    """Delete a fuel pump delivery (soft delete)."""
    fuel_pump_delivery_service.delete(db, delivery_id)

File: frontend/src/lib/api/fuel-pump-deliveries.ts
/**
 * Fuel Pump Deliveries API Service
 */

import type { FuelPumpDelivery, FuelPumpDeliverySummary } from "@/types";
import { apiClient } from "./client";
import type {
  FuelPumpDeliveryApiResponse,
  FuelPumpDeliveryCreateRequest,
  FuelPumpDeliveryUpdateRequest,
  FuelPumpDeliverySummaryResponse,
  ListParams,
} from "./types";
import {
  transformFuelPumpDelivery,
  transformFuelPumpDeliveries,
  transformFuelPumpDeliverySummary,
} from "./transformers";

export interface FuelPumpDeliveryListParams extends ListParams {
  pump_id?: string;
  start_date?: string;
  end_date?: string;
}

export const fuelPumpDeliveriesService = {
  /**
   * Get all fuel pump deliveries
   */
  async getAll(params?: FuelPumpDeliveryListParams): Promise<FuelPumpDelivery[]> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      skip: params?.skip,
      limit: params?.limit,
      pump_id: params?.pump_id,
      start_date: params?.start_date,
      end_date: params?.end_date,
    };

    const response = await apiClient.get<FuelPumpDeliveryApiResponse[]>(
      "/fuel-pump-deliveries",
      { params: queryParams }
    );
    return transformFuelPumpDeliveries(response);
  },

  /**
   * Get fuel pump delivery by ID
   */
  async getById(id: string): Promise<FuelPumpDelivery> {
    const response = await apiClient.get<FuelPumpDeliveryApiResponse>(
      `/fuel-pump-deliveries/${id}`
    );
    return transformFuelPumpDelivery(response);
  },

  /**
   * Get deliveries for a specific pump
   */
  async getByPump(pumpId: string, params?: ListParams): Promise<FuelPumpDelivery[]> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      skip: params?.skip,
      limit: params?.limit,
    };

    const response = await apiClient.get<FuelPumpDeliveryApiResponse[]>(
      `/fuel-pump-deliveries/pump/${pumpId}`,
      { params: queryParams }
    );
    return transformFuelPumpDeliveries(response);
  },

  /**
   * Get delivery summary for a date range
   */
  async getSummary(startDate: string, endDate: string): Promise<FuelPumpDeliverySummary> {
    const response = await apiClient.get<FuelPumpDeliverySummaryResponse>(
      "/fuel-pump-deliveries/summary",
      { params: { start_date: startDate, end_date: endDate } }
    );
    return transformFuelPumpDeliverySummary(response);
  },

  /**
   * Create new fuel pump delivery (automatically updates pump level)
   */
  async create(data: FuelPumpDeliveryCreateRequest): Promise<FuelPumpDelivery> {
    const response = await apiClient.post<FuelPumpDeliveryApiResponse>(
      "/fuel-pump-deliveries",
      data
    );
    return transformFuelPumpDelivery(response);
  },

  /**
   * Update fuel pump delivery
   */
  async update(id: string, data: FuelPumpDeliveryUpdateRequest): Promise<FuelPumpDelivery> {
    const response = await apiClient.patch<FuelPumpDeliveryApiResponse>(
      `/fuel-pump-deliveries/${id}`,
      data
    );
    return transformFuelPumpDelivery(response);
  },

  /**
   * Delete fuel pump delivery (soft delete)
   */
  async delete(id: string): Promise<void> {
    await apiClient.delete(`/fuel-pump-deliveries/${id}`);
  },
};
